Added WebServer DI sample

torbacz opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
- New WebServer sample with DI support in controllers.

## How Has This Been Tested?<!-- (if applicable) -->
<!--- Please describe in detail how you tested your sample/changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Deployed to ESP32 dev board and checked endpoint if working.

## Screenshots<!-- (if appropriate): -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Improvement (non-breaking change that improves a sample)
- [ ] Bug fix (fixes an issue with a current sample)
- [x] New Sample (adds a new sample)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- It would be nice if you could sign off your contribution by replacing the name with your GitHub user name and GitHub email contact. -->
Signed-off-by: GITHUB_USER <GITHUB_USER_EMAIL>

Ellerbach commented:
@torbacz we moved samples to the sample repository. Would you mind moving this one as well to the sample repository?

Ellerbach commented on samples/Webserver/WebServer.DI/webserver-cert.crt:
```diff
@@ -0,0 +1,20 @@
+-----BEGIN CERTIFICATE-----
```
you've checked in this cert but you're not using it. I guess it's because you've been testing it with https but in your sample, you're using http. So either use it, either remove it

torbacz replied:
Actually, I copied all files from other sample and forgot to delete it. Removed.

Ellerbach commented on samples/Webserver/WebServer.DI/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,33 @@
+﻿using System.Reflection;
```
maybe you can clean this one a bit

torbacz replied:
done

Ellerbach commented on samples/Webserver/WebServer.DI/TextServiceSingleton.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace nanoFramework.WebServer.Sample
+{
+    internal class TextServiceSingleton : ITextServiceSingleton
```
why do create 2 different classes which are exactly the same. Same for the interfaces. Only 1 is enough in both cases.
The difference is just when you inject it, either as a singleton either not. And I think it will be more clear as well

torbacz replied:
The idea behind two same interfaces and classes is very simple - I just want to use both in the same controller and show that one is new one each time and second is the same across all requests. And there is no way to do it in DI, because if you register same interface two times only the second one will be resolved.

![image](https://user-images.githubusercontent.com/19893965/187675242-8a55b955-26b2-488b-be06-b98e513152b6.png)
![image](https://user-images.githubusercontent.com/19893965/187675268-c9f34b5e-46aa-4e87-88af-49d316b2a839.png)

Hope it's clear now why, two different interfaces are required :)


I've cleaned it a little bit, by removing "TextServiceSingleton" class.

Ellerbach replied:
You can do it in a clear way as well with only 1 interface:

```csharp
public class ControllerTest
    {
        private readonly ITextService _textService;
        private readonly ITextService _textServiceSingleton;

        public ControllerTest(ITextService textService, ITextService textServiceSingleton)
        {
            _textService = textService;
            _textServiceSingleton = textServiceSingleton;
        }

        [Route("test")]
        [Method("GET")]
        public void RoutePostTest(WebServerEventArgs e)
        {
            var content = $"Response from {nameof(ITextService)}: {_textService.GetText()}. Response from {nameof(ITextServiceSingleton)}: {_textServiceSingleton.GetText()}";
            e.Context.Response.ContentType = "text/plain";
            WebServer.OutPutStream(e.Context.Response, content);
        }
    }
```

And also:

```csharp
private static ServiceProvider ConfigureServices()
        {
            return new ServiceCollection()
                .AddTransient(typeof(ITextService), typeof(TextService))
                .AddSingleton(typeof(ITextService), typeof(TextService))
                .BuildServiceProvider();
        }
```

You can fully get rid of the second interface and all will work the exact same way ;-)

The difference is only how you add the dependecies in this ServiceProvider function. One is added as a transient (so a new one everytime), the second one as a singleton (so unique).

give it a try ;-)

Ellerbach replied:
> And there is no way to do it in DI, because if you register same interface two times only the second one will be resolved

ho, I see now. So you're most likely right then!

torbacz replied:
![image](https://user-images.githubusercontent.com/19893965/187772076-47757b0c-4175-4dbf-b9af-c580701331f1.png)
Checked that, last one is resolved. So 2 interfaces and 1 class :)

Ellerbach reviewed (commented):
Nice :-) Thanks for adding. I would just simplify your interfaces and classes, only 1 is enough. The injection is what makes the difference in behavior

torbacz commented:
> @torbacz please address as well the issue with the other WebServer sample. It's failing in the build. Should be easy to fix.

Not sure what's wrong, with latest changes the build is working locally
![image](https://user-images.githubusercontent.com/19893965/187797539-71db97e6-13f3-49af-8527-9df699b651cf.png)

I'll try again later....

torbacz commented:
@Ellerbach 
"Good" news. I've got the same error locally.
![image](https://user-images.githubusercontent.com/19893965/188004140-27aa13e2-87fd-4bcb-9d30-8ca57a7dfae0.png)

Things, I've done to reproduce:
- Updated VS 2022 to latest (resulted in using msbuild 17.3.x instead of 17.2.x) - still no error
- Updated nanoFramework extension to 2022.2.0.37 - which results in error during build

Looks like there is something wrong with latest nanoFramework extension and generating resx files. To be honest, I've never worked with VS extensions, so it may take few days for me to fix it (any tips? )

BTW "easy to fix" 🤣

josesimoes commented:
I can confirm this. It's related with the ICO resource type.
This was brought in the latest code rework for the resx generator. The author is supposed to take care of it.
In the mean time we're OK with merging this PR.

No need for you to take care of this. It's a really complex tool to debug. Unless you want to, of course. 😅
